src/strands/tools/registry.py

```diff
@@ -24,244 +24,318 @@
-                self.load_tool_from_filepath(tool_name=tool["name"], tool_path=tool["path"])
-                tool_names.append(tool["name"])
-
-            # Case 3: Dictionary with path only
-            elif isinstance(tool, dict) and "path" in tool:
-                tool_name = os.path.basename(tool["path"]).split(".")[0]
-                self.load_tool_from_filepath(tool_name=tool_name, tool_path=tool["path"])
-                tool_names.append(tool_name)
-
-            # Case 4: Imported Python module
-            elif hasattr(tool, "__file__") and inspect.ismodule(tool):
-                # Get the module file path
-                module_path = tool.__file__
-                # Extract the tool name from the module name
-                tool_name = tool.__name__.split(".")[-1]
-
-                # Check for TOOL_SPEC in module to validate it's a Strands tool
-                if hasattr(tool, "TOOL_SPEC") and hasattr(tool, tool_name) and module_path:
-                    self.load_tool_from_filepath(tool_name=tool_name, tool_path=module_path)
-                    tool_names.append(tool_name)
-                else:
-                    function_tools = self._scan_module_for_tools(tool)
-                    for function_tool in function_tools:
-                        self.register_tool(function_tool)
-                        tool_names.append(function_tool.tool_name)
+        # Extract the tool name to check if it already exists
+        tool_name = self._extract_tool_name(tool)
+        if not tool_name:
+            raise ValueError("Could not extract tool name from provided tool")
 
-                    if not function_tools:
-                        logger.warning("tool_name=<%s>, module_path=<%s> | invalid agent tool", tool_name, module_path)
+        # Check if the tool already exists before making any changes
+        if tool_name in self.agent_tools:
+            raise ValueError(f"Tool '{tool_name}' already exists")
 
-            # Case 5: AgentTools (which also covers @tool)
-            elif isinstance(tool, AgentTool):
-                self.register_tool(tool)
-                tool_names.append(tool.tool_name)
-            else:
-                logger.warning("tool=<%s> | unrecognized tool specification", tool)
+        # Process the tool - this will validate and register the tool
+        created_tool_name = self._process_tool(tool, validate_unique=True)
 
-        return tool_names
+        if not created_tool_name:
+            raise ValueError(f"Failed to create tool '{tool_name}'")
 
-    def load_tool_from_filepath(self, tool_name: str, tool_path: str) -> None:
-        """Load a tool from a file path.
+        # Update tool configuration
+        if created_tool_name in self.agent_tools:
+            tool_spec = self.agent_tools[created_tool_name].tool_spec
+            self._update_tool_config(self._tool_config, tool_spec)
+
+        return created_tool_name
+
+    def read_tool(self, tool_name: str) -> ToolSpec:
```

Comment:
Per the description, this seems more like `read_tool_spec` instead of `read_tool`